File: m03_robot_description/m03_tf2_with_py/m03_tf2_with_py/lettuce_stick_frame.py
# !/usr/bin/env python3

# -------------------- Import Python Standard Headers -------------------------
import math

# ------------------- Import ROS2 rclpy implementations -----------------------
import rclpy
from rclpy.node import Node

# ----------------------- Import tf2 related packages -------------------------
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster

# ----------------------- Dynamic frame broadcaster class ---------------------
class LettuceStickBroadcaster(Node):
    """
    Class oriented to create a dynamic (moving) frame, that follows the idea of
    a lettuce on a stick (carrot on stick case from minecraft), as it will act 
    as an objective for others transforms to follow.

    Attributes
    ---
    tf_broadcaster : Transform Broadcaster
        Form to communicate the dynamic transform of interest.
        
    timer : Timer
        Timer that will call an update of the tf values.
        
    Methods
    ---
    broadcast_timer_callback:
        Linked to the timer call for dynamic update of the transform.
    """
    def __init__(self):
        """
        'Constructor that initialize a node called 
        'lettuce_stick_tf2_broadcaster', instance a transform broadcaster and
        then instance a timer linked to tf updates.
        """
        # Initialize node
        super().__init__('lettuce_stick_tf2_broadcaster')
        
        # Intance tf2 broadcaster
        self.tf_broadcaster = TransformBroadcaster(self)
        
        # Instance timer and link callback
        self.timer = self.create_timer(0.25, self.broadcast_timer_callback)

    def broadcast_timer_callback(self):
        """
        Publish a dynamic trnasform that will consider the time as the
        variation objetive, it will try to mimic the vibration of a lettuce
        when it is hanging on a stick that is attached to the animal.
        """
        # Time considerations
        seconds, _ = self.get_clock().now().seconds_nanoseconds()
        x = seconds * math.pi

        # Instance transform and consider stamped info
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'turtle1'
        t.child_frame_id = 'lettuce'
        
        # Update properly the 2D transform values
        t.transform.translation.x = 10 * math.sin(x)
        t.transform.translation.y = 10 * math.cos(x)
        t.transform.translation.z = 0.0
        t.transform.rotation.x = 0.0
        t.transform.rotation.y = 0.0
        t.transform.rotation.z = 0.0
        t.transform.rotation.w = 1.0

        # Publish (broadcast) transform
        self.tf_broadcaster.sendTransform(t)


def main():
    """
    Program oriented to create a dynamic transform to be followed. In this case
    an analogy of a lettuce attached to a stick.
    """
    
    # Initialize ROS Client Library for Python
    rclpy.init()
    
    # Create a logger
    logger = rclpy.logging.get_logger('logger')
    logger.info("Initializing lettuce and stick dynamic frame....")
    
    # Instance a node
    node = LettuceStickBroadcaster()
    
    # Spin node until keyboard interrupt happens
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    # Close and exit program
    logger.info("Closing lettuce and stick dynamic frame....")
    rclpy.shutdown()

File: m03_robot_description/m03_tf2_with_cpp/src/lettuce_stick_frame.cpp
// --------------------  STANDARD HEADERS REQUIRED -----------------------------
#include <chrono>      // Time mangement with different precisions
#include <functional>  // For function usage and hash related
#include <memory>      // Dynamic memory management
#include <cmath>       // C constants from math and operations 

// ------------------------ ROS2 RELATED HEADERS ------------------------------
#include "rclcpp/rclcpp.hpp"                // ROS2 Client Library for C++
#include "tf2_ros/transform_broadcaster.h"  // TF broadcaster

// ------------------------ ROS2 MESSAGES -------------------------------------
#include "geometry_msgs/msg/transform_stamped.hpp" // Tf with time stamp

// ------------------------ NAMESPACE CONSIDERATIONS --------------------------
using namespace std::chrono_literals;  // For user-defined chrono literals

// ----------------- DYNAMIC FRAME BROADCASTER IMPLEMENTATION -----------------

/**
 * Class that publishes a dynamic frame from the turtle, you can understand it
 * as it was a firshing rot with a carrot with a lot of wind, a very long
 * stick and a very long rope, this is made by using sine and cosine functions
 */
class LettuceStickBroadcaster : public rclcpp::Node
{
public:
    /**
     * User defined constructor that initialize the node under the name 
     * "lettue_stick_broadcaster", declare a broadcaster and a timer linked
     * with a callback to publish the tf.
     */
    LettuceStickBroadcaster()
        : Node("lettuce_stick_broadcaster")
    {
        // Declare a transform broadcater
        tf_broad_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        
        // Timer that will be linked with the publish and broadcast fo the tf
        timer_ = this->create_wall_timer(
            100ms, std::bind(&LettuceStickBroadcaster::broad_callback, this));
    }
private:
    /**
     * Callback that takes place when a timer request it, to broadcast a
     * tranform that is dynamically related to the turtle and circle 
     * identitites
     */
    void broad_callback()
    {
        // Get current time and process it with pi
        rclcpp::Time now = this->get_clock()->now();
        double x = now.seconds() * M_PI;

        // Declare stamped transform
        geometry_msgs::msg::TransformStamped t_stamp;

        // Add header information: Time parent and child
        t_stamp.header.stamp = now;
        t_stamp.header.frame_id = "turtle1";
        t_stamp.child_frame_id = "lettuce1";

        // X and Y traslational values related with sine and cosine of time
        t_stamp.transform.translation.x = 10 * sin(x);
        t_stamp.transform.translation.y = 10 * cos(x);
        t_stamp.transform.translation.z = 0.0;

        // Rotation remains const
        t_stamp.transform.rotation.x = 0.0;
        t_stamp.transform.rotation.y = 0.0;
        t_stamp.transform.rotation.z = 0.0;
        t_stamp.transform.rotation.w = 1.0;

        // Publish transform
        tf_broad_->sendTransform(t_stamp);
    }

    // Declare timer and transfor broadcaster
    rclcpp::TimerBase::SharedPtr timer_;
    std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broad_;
};

int main(int argc, char * argv[])
{
    // Initialize ROS2 Cleint Library for C++
    rclcpp::init(argc, argv);

    // Spin node creation
    rclcpp::spin(std::make_shared<LettuceStickBroadcaster>());

    // When spin is interrupted, close and shutdown
    rclcpp::shutdown();
    return 0;
}